File: packages/js-core/src/queries/query-manager.ts
import { CommentsFeedState, ModerationFilters, ServerError } from '../utils';
import type { PolycentricClient } from '../polycentric-client';
import {
  ContentType,
  LWWElement,
  Pointer,
  PublicKey,
  Event as ProtobufEvent,
  FeedResult,
  ImageManifest,
  Events,
  ResultEventsAndRelatedEventsAndCursor,
  Reference,
  EventKey,
} from '../proto/polycentric';
import { FeedQuery } from './feed-query';

interface ResultAndServerErrors {
  result: Uint8Array;
  errors: ServerError[];
}

export class QueryManager {
  constructor(private readonly client: PolycentricClient) {}

  queryExploreFeed(
    perServerLimit?: number,
    moderationFilters?: ModerationFilters,
  ): FeedQuery {
    const currentSystem = this.client.currentIdentity.keyPair.publicKey;
    const currentSystemBytes = PublicKey.toBinary(currentSystem);

    const filters = moderationFilters
      ? JSON.stringify(moderationFilters)
      : null;
    const getExploreCallback = this.client.httpClient.getExplore.bind(
      this.client.httpClient,
    );

    return new FeedQuery(this.client, async (cursors, _latestEvent) => {
      const result = (await this.client.wasmCore.query_explore_feed(
        currentSystemBytes,
        getExploreCallback,
        cursors,
        perServerLimit,
        filters,
      )) as ResultAndServerErrors;

      return {
        events: Events.fromBinary(result.result),
        errors: result.errors,
      };
    });
  }

  queryExploreFeedSpecificServer(
    server: string,
    limit?: number,
    moderationFilters?: ModerationFilters,
  ): FeedQuery {
    const filters = moderationFilters
      ? JSON.stringify(moderationFilters)
      : null;
    const getExploreCallback = this.client.httpClient.getExplore.bind(
      this.client.httpClient,
    );

    return new FeedQuery(this.client, async (cursors, _latestEvent) => {
      const response =
        await this.client.wasmCore.query_explore_feed_specific_server(
          server,
          getExploreCallback,
          cursors.get(server),
          limit,
          filters,
        );

      const result = ResultEventsAndRelatedEventsAndCursor.fromBinary(response);

      if (result.cursor) cursors.set(server, result.cursor);

      return {
        events: result.resultEvents || { events: [] },
        errors: [],
      };
    });
  }

  querySearch(
    searchQuery: string,
    searchType?: 'messages' | 'profiles',
    perServerLimit?: number,
    moderationFilters?: ModerationFilters,
  ): FeedQuery {
    const currentSystem = this.client.currentIdentity.keyPair.publicKey;
    const currentSystemBytes = PublicKey.toBinary(currentSystem);

    const getSearchCallback = this.client.httpClient.getSearch.bind(
      this.client.httpClient,
    );

    const filters = moderationFilters
      ? JSON.stringify(moderationFilters)
      : null;

    return new FeedQuery(this.client, async (cursors, _latestEvent) => {
      const result = (await this.client.wasmCore.query_search(
        currentSystemBytes,
        getSearchCallback,
        searchQuery,
        searchType,
        cursors,
        perServerLimit,
        filters,
      )) as ResultAndServerErrors;

      return {
        events: Events.fromBinary(result.result),
        errors: result.errors,
      };
    });
  }

  queryFollowingFeed(limit: number): FeedQuery {
    const currentSystem = this.client.currentIdentity.keyPair.publicKey;
    const currentSystemBytes = PublicKey.toBinary(currentSystem);

    return new FeedQuery(this.client, async (_cursors, latestEvent) => {
      const latestEventBytes = latestEvent
        ? ProtobufEvent.toBinary(latestEvent)
        : undefined;

      const result = this.client.wasmCore.query_following_feed(
        currentSystemBytes,
        limit,
        latestEventBytes,
      );

      return {
        events: Events.fromBinary(result),
        errors: [],
      };
    });
  }

  queryAuthorFeed(profile: PublicKey, limit: number): FeedQuery {
    const currentSystem = this.client.currentIdentity.keyPair.publicKey;
    const currentSystemBytes = PublicKey.toBinary(currentSystem);

    const profileBytes = PublicKey.toBinary(profile);

    const getHeadCallback = this.client.httpClient.getHead.bind(
      this.client.httpClient,
    );
    const getRangesCallback = this.client.httpClient.getRanges.bind(
      this.client.httpClient,
    );
    const getEventsCallback = this.client.httpClient.getEvents.bind(
      this.client.httpClient,
    );

    return new FeedQuery(this.client, async (_cursors, latestEvent) => {
      const latestEventBytes = latestEvent
        ? ProtobufEvent.toBinary(latestEvent)
        : undefined;

      const result = await this.client.wasmCore.query_author_feed(
        currentSystemBytes,
        profileBytes,
        limit,
        latestEventBytes,
        getHeadCallback,
        getRangesCallback,
        getEventsCallback,
      );

      return {
        events: Events.fromBinary(result),
        errors: [],
      };
    });
  }

  queryReferencesFeed(
    reference: Reference,
    moderationFilters?: ModerationFilters,
  ): FeedQuery {
    const currentSystem = this.client.currentIdentity.keyPair.publicKey;
    const currentSystemBytes = PublicKey.toBinary(currentSystem);

    const getQueryReferencesCallback =
      this.client.httpClient.getQueryReferences.bind(this.client.httpClient);

    const referenceBytes = Reference.toBinary(reference);
    const filters = moderationFilters
      ? JSON.stringify(moderationFilters)
      : null;

    return new FeedQuery(this.client, async (cursors, _latestEvent) => {
      const result = await this.client.wasmCore.query_references_feed(
        currentSystemBytes,
        getQueryReferencesCallback,
        referenceBytes,
        cursors,
        filters,
      );

      return {
        events: Events.fromBinary(result.result),
        errors: result.errors,
      };
    });
  }

  queryLikesFeed(limit: number): FeedQuery {
    const currentSystem = this.client.currentIdentity.keyPair.publicKey;
    const currentSystemBytes = PublicKey.toBinary(currentSystem);

    return new FeedQuery(this.client, async (_cursors, latestEvent) => {
      const latestEventBytes = latestEvent
        ? ProtobufEvent.toBinary(latestEvent)
        : undefined;

      const result = this.client.wasmCore.query_likes_feed(
        currentSystemBytes,
        limit,
        latestEventBytes,
      );

      return {
        events: Events.fromBinary(result),
        errors: [],
      };
    });
  }

  queryCommentsFeed(moderationFilters?: ModerationFilters): FeedQuery {
    const currentSystem = this.client.currentIdentity.keyPair.publicKey;
    const currentSystemBytes = PublicKey.toBinary(currentSystem);

    const getQueryReferencesCallback =
      this.client.httpClient.getQueryReferences.bind(this.client.httpClient);

    const filters = moderationFilters
      ? JSON.stringify(moderationFilters)
      : null;

    const feedState: CommentsFeedState = {};

    return new FeedQuery(this.client, async (_cursors, _latestEvent) => {
      const result = await this.client.wasmCore.query_comments_feed(
        currentSystemBytes,
        getQueryReferencesCallback,
        feedState,
        filters,
      );

      return {
        events: Events.fromBinary(result.result),
        errors: result.errors,
      };
    });
  }

  queryCurrentOpinion(targetPointer: Pointer): LWWElement | null {
    const currentSystem = this.client.currentIdentity.keyPair.publicKey;
    const currentSystemBytes = PublicKey.toBinary(currentSystem);
    const targetPointerBytes = Pointer.toBinary(targetPointer);

    const result = this.client.wasmCore.query_opinion(
      currentSystemBytes,
      targetPointerBytes,
    );

    if (!result) return null;

    return LWWElement.fromBinary(result);
  }

  queryIsDeleted(targetPointer: Pointer): boolean {
    return this.client.wasmCore.query_event_is_deleted(
      Pointer.toBinary(targetPointer),
    );
  }

  queryFeed(
    system: PublicKey,
    options: {
      startTime?: bigint;
      endTime?: bigint;
      limit?: number;
      cursor?: Uint8Array;
    } = {},
  ): FeedResult {
    const systemBytes = PublicKey.toBinary(system);
    const { startTime, endTime, limit, cursor } = options;

    const result = this.client.wasmCore.query_feed_with_cursor(
      systemBytes,
      startTime,
      endTime,
      limit,
      cursor,
    );

    if (!result) {
      return FeedResult.create({
        events: [],
        cursor: new Uint8Array(),
      });
    }

    return FeedResult.fromBinary(result);
  }

  async queryUsername(system: PublicKey): Promise<string | null> {
    const lwwElement = await this._querySystemCRDT(
      ContentType.USERNAME,
      system,
    );
    if (!lwwElement) return null;
    return new TextDecoder().decode(lwwElement.value);
  }

  async queryDescription(system: PublicKey): Promise<string | null> {
    const lwwElement = await this._querySystemCRDT(
      ContentType.DESCRIPTION,
      system,
    );
    if (!lwwElement) return null;
    return new TextDecoder().decode(lwwElement.value);
  }

  async queryAvatar(system: PublicKey): Promise<ImageManifest | null> {
    const lwwElement = await this._querySystemCRDT(ContentType.AVATAR, system);
    if (!lwwElement) return null;
    return ImageManifest.fromBinary(lwwElement.value);
  }

  async queryBanner(system: PublicKey): Promise<ImageManifest | null> {
    const lwwElement = await this._querySystemCRDT(ContentType.BANNER, system);
    if (!lwwElement) return null;
    return ImageManifest.fromBinary(lwwElement.value);
  }

  queryFollows(system: PublicKey): PublicKey[] {
    const systemBytes = PublicKey.toBinary(system);
    const result = this.client.wasmCore.query_follows_for_system(systemBytes);

    if (!result) return [];

    const events = Events.fromBinary(result);
    return events.events
      .map((signedEvent) => {
        const event = ProtobufEvent.fromBinary(signedEvent.event);
        if (!event.lwwElementSet) return undefined;
        return PublicKey.fromBinary(event.lwwElementSet.value);
      })
      .filter(Boolean) as PublicKey[];
  }

  queryBlocks(system: PublicKey): PublicKey[] {
    const systemBytes = PublicKey.toBinary(system);
    const result = this.client.wasmCore.query_blocks_for_system(systemBytes);

    if (!result) return [];

    const events = Events.fromBinary(result);
    return events.events
      .map((signedEvent) => {
        const event = ProtobufEvent.fromBinary(signedEvent.event);
        if (!event.lwwElementSet) return undefined;
        return PublicKey.fromBinary(event.lwwElementSet.value);
      })
      .filter(Boolean) as PublicKey[];
  }

  queryServers(system: PublicKey): string[] {
    const systemBytes = PublicKey.toBinary(system);
    const result = this.client.wasmCore.query_servers_for_system(systemBytes);

    if (!result) return [];

    const events = Events.fromBinary(result);
    return events.events
      .map((signedEvent) => {
        const event = ProtobufEvent.fromBinary(signedEvent.event);
        if (!event.lwwElementSet) return undefined;
        return new TextDecoder().decode(event.lwwElementSet.value);
      })
      .filter(Boolean) as string[];
  }

  queryAuthorities(system: PublicKey): string[] {
    const systemBytes = PublicKey.toBinary(system);
    const result =
      this.client.wasmCore.query_authorities_for_system(systemBytes);

    if (!result) return [];

    const events = Events.fromBinary(result);
    return events.events
      .map((signedEvent) => {
        const event = ProtobufEvent.fromBinary(signedEvent.event);
        if (!event.lwwElementSet) return undefined;
        return new TextDecoder().decode(event.lwwElementSet.value);
      })
      .filter(Boolean) as string[];
  }

  queryTopics(system: PublicKey): string[] {
    const systemBytes = PublicKey.toBinary(system);
    const result = this.client.wasmCore.query_topics_for_system(systemBytes);

    if (!result) return [];

    const events = Events.fromBinary(result);
    return events.events
      .map((signedEvent) => {
        const event = ProtobufEvent.fromBinary(signedEvent.event);
        if (!event.lwwElementSet) return undefined;
        return new TextDecoder().decode(event.lwwElementSet.value);
      })
      .filter(Boolean) as string[];
  }

  private async _querySystemCRDT(
    contentType: ContentType,
    system: PublicKey,
  ): Promise<LWWElement | null> {
    const currentSystem = this.client.currentIdentity.keyPair.publicKey;
    const currentSystemBytes = PublicKey.toBinary(currentSystem);

    const systemBytes = PublicKey.toBinary(system);
    const result = await this.client.wasmCore.query_crdt_for_system(
      systemBytes,
      BigInt(contentType),
      currentSystemBytes,
      this.client.httpClient.getQueryLatest.bind(this.client.httpClient),
    );

    if (!result) return null;

    return LWWElement.fromBinary(result);
  }

  eventPointer(event: ProtobufEvent): Pointer {
    const eventBytes = ProtobufEvent.toBinary(event);
    const pointerBytes = this.client.wasmCore.get_pointer(eventBytes);
    return Pointer.fromBinary(pointerBytes);
  }

  eventKey(event: ProtobufEvent): EventKey {
    const pointer = Pointer.toBinary(this.eventPointer(event));
    const eventKeyBytes = this.client.wasmCore.get_reference(pointer);

    if (!eventKeyBytes) throw new TypeError('Event is missing required fields');

    return EventKey.fromBinary(eventKeyBytes);
  }
}

File: packages/kotlin/app/src/main/java/tech/futo/libPolycentric/queries/QueryManager.kt
package tech.futo.libPolycentric.queries

import okio.ByteString
import okio.ByteString.Companion.toByteString
import polycentric.ContentType
import polycentric_ffi.Option
import polycentric.Event
import polycentric.EventKey
import polycentric.Events
import polycentric.FeedResult
import polycentric.ImageManifest
import polycentric.LWWElement
import polycentric.Pointer
import polycentric.PublicKey
import polycentric.Reference
import polycentric_ffi.FeedQuery as FFIFeedQuery
import polycentric_ffi.SearchQuery
import polycentric_ffi.SearchType
import polycentric_ffi.ServerFeedQuery
import tech.futo.libPolycentric.PolycentricClient

class QueryManager(private val client: PolycentricClient) {

    fun queryExploreFeed(
        perServerLimit: Long? = null,
        moderationFilters: String? = null,
    ): FeedQuery {
        val currentSystemBytes = PublicKey.ADAPTER.encode(
            client.currentIdentity.keyPair.publicKey
        )

        val feedQuery = ServerFeedQuery(
            per_server_limit = perServerLimit,
            moderation_filters = moderationFilters,
        )
        val feedQueryBytes = ServerFeedQuery.ADAPTER.encode(feedQuery)

        return FeedQuery(client) { cursor ->
            client.ffiService.queryExploreFeed(currentSystemBytes, feedQueryBytes, cursor)
        }
    }

    fun querySearchFeed(
        searchQuery: String,
        searchType: String? = null,
        perServerLimit: Long? = null,
        moderationFilters: String? = null,
    ): FeedQuery {
        val currentSystemBytes = PublicKey.ADAPTER.encode(
            client.currentIdentity.keyPair.publicKey
        )

        val feedQuery = ServerFeedQuery(
            per_server_limit = perServerLimit,
            moderation_filters = moderationFilters,
        )
        val feedQueryBytes = ServerFeedQuery.ADAPTER.encode(feedQuery)

        val search = SearchQuery(
            query = searchQuery,
            type = when (searchType) {
                "profiles" -> SearchType.profiles
                else -> SearchType.messages
            },
        )
        val searchQueryBytes = SearchQuery.ADAPTER.encode(search)

        return FeedQuery(client) { cursor ->
            client.ffiService.querySearchFeed(currentSystemBytes, feedQueryBytes, searchQueryBytes, cursor)
        }
    }

    fun queryFollowingFeed(limit: Int): FeedQuery {
        val currentSystemBytes = PublicKey.ADAPTER.encode(
            client.currentIdentity.keyPair.publicKey
        )

        return FeedQuery(client) { cursor ->
            client.ffiService.queryFollowingFeed(currentSystemBytes, limit, cursor)
        }
    }

    fun queryAuthorFeed(profile: PublicKey, limit: Int): FeedQuery {
        val currentSystemBytes = PublicKey.ADAPTER.encode(
            client.currentIdentity.keyPair.publicKey
        )
        val profileBytes = PublicKey.ADAPTER.encode(profile)

        return FeedQuery(client) { cursor ->
            client.ffiService.queryAuthorFeed(currentSystemBytes, profileBytes, limit, cursor)
        }
    }

    fun queryReferencesFeed(
        reference: Reference,
        moderationFilters: String? = null,
    ): FeedQuery {
        val currentSystemBytes = PublicKey.ADAPTER.encode(
            client.currentIdentity.keyPair.publicKey
        )

        val feedQuery = ServerFeedQuery(
            moderation_filters = moderationFilters,
        )
        val feedQueryBytes = ServerFeedQuery.ADAPTER.encode(feedQuery)
        val referenceBytes = Reference.ADAPTER.encode(reference)

        return FeedQuery(client) { cursor ->
            client.ffiService.queryReferencesFeed(currentSystemBytes, feedQueryBytes, referenceBytes, cursor)
        }
    }

    fun queryLikesFeed(limit: Int): FeedQuery {
        val currentSystemBytes = PublicKey.ADAPTER.encode(
            client.currentIdentity.keyPair.publicKey
        )

        return FeedQuery(client) { cursor ->
            client.ffiService.queryLikesFeed(currentSystemBytes, limit, cursor)
        }
    }

    fun queryCommentsFeed(moderationFilters: String? = null): FeedQuery {
        val currentSystemBytes = PublicKey.ADAPTER.encode(
            client.currentIdentity.keyPair.publicKey
        )

        val feedQuery = ServerFeedQuery(
            moderation_filters = moderationFilters,
        )
        val feedQueryBytes = ServerFeedQuery.ADAPTER.encode(feedQuery)

        return FeedQuery(client) { cursor ->
            client.ffiService.queryCommentsFeed(currentSystemBytes, feedQueryBytes, cursor)
        }
    }

    suspend fun queryCurrentOpinion(targetPointer: Pointer): LWWElement? {
        val currentSystemBytes = PublicKey.ADAPTER.encode(
            client.currentIdentity.keyPair.publicKey
        )
        val targetPointerBytes = Pointer.ADAPTER.encode(targetPointer)

        val result = client.ffiService.queryOpinion(currentSystemBytes, targetPointerBytes)
        if (result.isEmpty()) return null

        val option = Option.ADAPTER.decode(result)
        val valueBytes = option.value_ ?: return null

        return LWWElement.ADAPTER.decode(valueBytes.toByteArray())
    }

    suspend fun queryIsDeleted(targetPointer: Pointer): Boolean {
        val result = client.ffiService.queryEventIsDeleted(
            Pointer.ADAPTER.encode(targetPointer)
        )
        return result.isNotEmpty() && result[0] == 1.toByte()
    }

    suspend fun queryFeed(
        system: PublicKey,
        startTime: Long? = null,
        endTime: Long? = null,
        limit: Long? = null,
        cursor: ByteArray? = null,
    ): FeedResult {
        val feedQuery = FFIFeedQuery(
            system_bytes = PublicKey.ADAPTER.encode(system).toByteString(),
            start_time = startTime,
            end_time = endTime,
            limit = limit,
            cursor = cursor?.toByteString(),
        )
        val feedQueryBytes = FFIFeedQuery.ADAPTER.encode(feedQuery)

        val result = client.ffiService.queryFeedWithCursor(feedQueryBytes)
        if (result.isEmpty()) {
            return FeedResult(events = emptyList(), cursor = ByteString.EMPTY)
        }
        return FeedResult.ADAPTER.decode(result)
    }

    suspend fun queryUsername(system: PublicKey): String? {
        val lwwElement = querySystemCRDT(ContentType.USERNAME, system) ?: return null
        return lwwElement.value_.toByteArray().decodeToString()
    }

    suspend fun queryDescription(system: PublicKey): String? {
        val lwwElement = querySystemCRDT(ContentType.DESCRIPTION, system) ?: return null
        return lwwElement.value_.toByteArray().decodeToString()
    }

    suspend fun queryAvatar(system: PublicKey): ImageManifest? {
        val lwwElement = querySystemCRDT(ContentType.AVATAR, system) ?: return null
        return ImageManifest.ADAPTER.decode(lwwElement.value_.toByteArray())
    }

    suspend fun queryBanner(system: PublicKey): ImageManifest? {
        val lwwElement = querySystemCRDT(ContentType.BANNER, system) ?: return null
        return ImageManifest.ADAPTER.decode(lwwElement.value_.toByteArray())
    }

    suspend fun queryFollows(system: PublicKey): List<PublicKey> {
        val systemBytes = PublicKey.ADAPTER.encode(system)
        val result = client.ffiService.queryFollowsForSystem(systemBytes)
        if (result.isEmpty()) return emptyList()

        return extractPublicKeysFromLWWSetEvents(result)
    }

    suspend fun queryBlocks(system: PublicKey): List<PublicKey> {
        val systemBytes = PublicKey.ADAPTER.encode(system)
        val result = client.ffiService.queryBlocksForSystem(systemBytes)
        if (result.isEmpty()) return emptyList()

        return extractPublicKeysFromLWWSetEvents(result)
    }

    suspend fun queryServers(system: PublicKey): List<String> {
        val systemBytes = PublicKey.ADAPTER.encode(system)
        val result = client.ffiService.queryServersForSystem(systemBytes)
        if (result.isEmpty()) return emptyList()

        return extractStringsFromLWWSetEvents(result)
    }

    suspend fun queryAuthorities(system: PublicKey): List<String> {
        val systemBytes = PublicKey.ADAPTER.encode(system)
        val result = client.ffiService.queryAuthoritiesForSystem(systemBytes)
        if (result.isEmpty()) return emptyList()

        return extractStringsFromLWWSetEvents(result)
    }

    suspend fun queryTopics(system: PublicKey): List<String> {
        val systemBytes = PublicKey.ADAPTER.encode(system)
        val result = client.ffiService.queryTopicsForSystem(systemBytes)
        if (result.isEmpty()) return emptyList()

        return extractStringsFromLWWSetEvents(result)
    }

    suspend fun eventPointer(event: Event): Pointer {
        val eventBytes = Event.ADAPTER.encode(event)
        val pointerBytes = client.ffiService.getPointer(eventBytes)
        return Pointer.ADAPTER.decode(pointerBytes)
    }

    suspend fun eventKey(event: Event): EventKey {
        val pointer = eventPointer(event)
        val pointerBytes = Pointer.ADAPTER.encode(pointer)
        val eventKeyBytes = client.ffiService.getReference(pointerBytes)
        require(eventKeyBytes.isNotEmpty()) { "Event is missing required fields" }
        return EventKey.ADAPTER.decode(eventKeyBytes)
    }

    private suspend fun querySystemCRDT(contentType: ContentType, system: PublicKey): LWWElement? {
        val currentSystemBytes = PublicKey.ADAPTER.encode(
            client.currentIdentity.keyPair.publicKey
        )
        val systemBytes = PublicKey.ADAPTER.encode(system)

        val result = client.ffiService.queryCrdtForSystem(
            systemBytes, contentType.value, currentSystemBytes,
        )
        if (result.isEmpty()) return null

        val option = Option.ADAPTER.decode(result)
        val valueBytes = option.value_ ?: return null

        return LWWElement.ADAPTER.decode(valueBytes.toByteArray())
    }

    private fun extractPublicKeysFromLWWSetEvents(resultBytes: ByteArray): List<PublicKey> {
        val events = Events.ADAPTER.decode(resultBytes)
        return events.events.mapNotNull { signedEvent ->
            val event = Event.ADAPTER.decode(signedEvent.event.toByteArray())
            val lwwElementSet = event.lww_element_set ?: return@mapNotNull null
            PublicKey.ADAPTER.decode(lwwElementSet.value_.toByteArray())
        }
    }

    private fun extractStringsFromLWWSetEvents(resultBytes: ByteArray): List<String> {
        val events = Events.ADAPTER.decode(resultBytes)
        return events.events.mapNotNull { signedEvent ->
            val event = Event.ADAPTER.decode(signedEvent.event.toByteArray())
            val lwwElementSet = event.lww_element_set ?: return@mapNotNull null
            lwwElementSet.value_.toByteArray().decodeToString()
        }
    }

}
